main.go

```diff
@@ -196,19 +199,67 @@ func StartPurgeLoop(secInterval int) {
 	}
 }
 
-func writeToPumps(keys []interface{}, job *health.Job, startTime time.Time) {
+func writeToPumps(keys []interface{}, job *health.Job, startTime time.Time, purgeDelay int) {
 	// Send to pumps
 	if Pumps != nil {
+		var wg sync.WaitGroup
+		wg.Add(len(Pumps))
+
 		for _, pmp := range Pumps {
-			log.WithFields(logrus.Fields{
-				"prefix": mainPrefix,
-			}).Debug("Writing to: ", pmp.GetName())
-			pmp.WriteData(keys)
-			if job != nil {
-				job.Timing("purge_time_"+pmp.GetName(), time.Since(startTime).Nanoseconds())
-			}
+			go func(wg *sync.WaitGroup, pmp pumps.Pump, keys *[]interface{}) {
+				timer := time.AfterFunc(time.Duration(purgeDelay)*time.Second, func() {
+					if pmp.GetTimeout() == 0 {
+						log.WithFields(logrus.Fields{
+							"prefix": mainPrefix,
+						}).Warning("Pump  ", pmp.GetName(), " is taking more time than the value configured of purge_delay. You should try to set a timeout for this pump.")
+					} else if pmp.GetTimeout() > purgeDelay {
+						log.WithFields(logrus.Fields{
+							"prefix": mainPrefix,
+						}).Warning("Pump  ", pmp.GetName(), " is taking more time than the value configured of purge_delay. You should try lowering the timeout configured for this pump.")
+					}
+				})
+				defer timer.Stop()
+				defer wg.Done()
+				ch := make(chan error, 1)
+
+				//Load pump timeout
+				timeout := pmp.GetTimeout()
+				var ctx context.Context
+				var cancel context.CancelFunc
+				//Initialize context depending if the pump has a configured timeout
+				if timeout > 0 {
+					ctx, cancel = context.WithTimeout(context.Background(), time.Duration(timeout)*time.Second)
+				} else {
+					ctx, cancel = context.WithCancel(context.Background())
+				}
+
+				defer cancel()
+
+				go func(ch chan error, ctx context.Context, pmp pumps.Pump, keys *[]interface{}) {
+					err := pmp.WriteData(ctx, *keys)
+					ch <- err
+				}(ch, ctx, pmp, keys)
+
+				select {
+				case err := <-ch:
+					if err != nil {
+						log.WithFields(logrus.Fields{
+							"prefix": mainPrefix,
+						}).Warning("Error Writing to: ", pmp.GetName(), " - Error:", err)
+					}
+				case <-ctx.Done():
+					log.WithFields(logrus.Fields{
+						"prefix": mainPrefix,
+					}).Warning("Timeout Writing to: ", pmp.GetName())
```

Comment:
Well, this will not only hold for timeouts, if the context was cancelled or any of the parent chain of the context were cancelled it will trigger this.

We need something like this 

```go
		switch ctx.Err() {
		case context.Canceled:
			// log something so we know we didn't timeout but the job was cancelled?
		case context.DeadlineExceeded:
			log.WithFields(logrus.Fields{
				"prefix": mainPrefix,
			}).Warning("Timeout Writing to: ", pmp.GetName())
		}
```